rotkehlchen/history/events/structures/asset_movement.py

```diff
@@ -0,0 +1,200 @@
+from rotkehlchen.accounting.structures.balance import Balance
+from rotkehlchen.assets.asset import Asset
+from rotkehlchen.constants import ZERO
+from rotkehlchen.constants.location_details import get_formatted_location_name
+from rotkehlchen.exchanges.data_structures import hash_id
+from rotkehlchen.history.events.structures.base import (
+    HISTORY_EVENT_DB_TUPLE_WRITE,
+    HistoryBaseEntry,
+    HistoryBaseEntryType,
+)
+from rotkehlchen.history.events.structures.types import (
+    HistoryEventSubType,
+    HistoryEventType,
+)
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.serialization.deserialize import deserialize_fval
+from rotkehlchen.types import (
+    Location,
+    TimestampMS,
+)
+from rotkehlchen.utils.misc import ts_ms_to_sec
+
+if TYPE_CHECKING:
+    from more_itertools import peekable
+
+    from rotkehlchen.accounting.mixins.event import AccountingEventMixin
+    from rotkehlchen.accounting.pot import AccountingPot
+    from rotkehlchen.fval import FVal
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class AssetMovement(HistoryBaseEntry):
+    """Asset movement event representing deposits and withdrawals on exchanges."""
+
+    def __init__(
+            self,
+            sequence_index: int,
+            timestamp: TimestampMS,
+            location: Location,
+            event_type: HistoryEventType,
+            event_subtype: HistoryEventSubType,
+            asset: Asset,
+            balance: Balance,
+            location_label: str | None = None,
+            notes: str | None = None,
+            identifier: int | None = None,
+            event_identifier: str | None = None,
+            extra_data: dict[str, Any] | None = None,
+            address: str | None = None,
+            transaction_id: str | None = None,
+            link: str | None = None,
+    ) -> None:
+        """
+        Optional args stored in extra_data:
+        `address`: The source address for deposits and the destination address for withdrawals.
+        `transaction_id`: Transaction id on the exchange or the associated onchain transaction hash.
+        `link`: For exchange asset movements this should be the exchange unique identifier.
+            For movements imported from third parties we should generate a unique id for this.
+            If movements are both imported from third parties like cointracking.info and from
+            the exchanges themselves then there is no way to avoid duplicates.
+        """  # noqa: E501
+        for key, value in {
+            'address': address,
+            'transaction_id': transaction_id,
+            'link': link,
+        }.items():
+            if value is None:
+                continue
+
+            if extra_data is None:
+                extra_data = {}
+
+            extra_data[key] = value
+
+        super().__init__(
+            event_identifier=event_identifier if event_identifier is not None else hash_id(
```

Comment:
so no this is not gonna work. This will re-create the same problem I posted about in the morning that we want to avoid: https://github.com/rotki/rotki/issues/3010

We can't have such unpredictable identifiers. We need to make sure that events never get dupicated and uniqueidentifiers are unique.

The how is a problem which you need to think about.

If there is a `unique_id` from the exchange use exchangename + uniqueid.
If not then you need to get creative.

And that creativeness should definitely be encoded in a function so it can be reused in other places where the unique id would need to be recreated.

Perhaps enough would be: location | timestamp | asset | amount.

Only problem then would be if you have somehow two duplicate deposit/withdrawals at the same milisecond for the same asset and the same amount. Which perhaps should indeed never happen?

Reply:
here it would be unique_id, maybe is good idea to have `location + unique_id` to prevent conflicts. Then I'm not sure what exchanges don't provide an id. Would need to check. Only thing that might be dangerous is imports from csvs where we need to check if we have ids for everything.

Reply:
Changed this to use location + unique_id when unique_id is present, and otherwise create it from location, timestamp, asset and amount as suggested.